Add `wxCursorBundle` to make supporting high resolution cursors simpler

vadz opened the pull request:
Closes #24564.

vadz committed:
Document that wxBitmap(wxCursor) doesn't work with Wayland

There doesn't seem to be any way to make it work, so at least document
that it doesn't.

vadz committed:
Update wxCursor description in the documentation

Avoid speaking of X and don't advise using conditional compilation when
cursors can be created portably (although the example still shows it).

vadz committed:
Add wxCursorBundle and implement support for it in wxGTK

This new class allows to let wxWidgets select the appropriate cursor,
similarly to how wxBitmapBundle allows to select the bitmap of the
appropriate size.

Document the new class and show using it in the drawing sample.

vadz committed:
Use wxSize in wxMSW wxCursor code instead of pair of ints

No real changes, just make the code a bit more readable by using a
single wxSize instead of a pair of ints.

vadz committed:
Also use wxPoint in wxMSW wxCursor implementation

Still no real changes, just make the code a bit shorter and less
repetitive.

vadz committed:
Cache the last returned wxCursor in wxCursorBundle

This is required in order to use the cursor returned by GetCursorFor()
with wxEVT_SET_CURSOR in wxMSW: without it, the cursor object itself is
destroyed too early and HCURSOR we return to the system from this event
handler becomes invalid.

It is also more efficient, which is important for a function which may
be called repeatedly on each mouse move.

vadz committed:
Stop rescaling wxImage before creating wxCursor from it

The restriction on the cursor sizes that this code tried to avoid
doesn't appear to exist any more: creating cursors of any size works
fine even without it under Windows 10.

So just stop doing this.

vadz committed:
Move check for cursor hot spot validity where it belongs

Check for this when actually creating the cursor, whatever we are
creating it from, instead of only checking this when creating cursor
from wxImage (but not wxBitmap).

vadz committed:
Remove outdated comment before wxSysColourChangedEvent declaration

Remove a wrong TODO comment dating from prehistorical wxWidgets version.

No real changes.

vadz committed:
Add wxSYS_CURSOR_SIZE

As wxSYS_CURSOR_X and wxSYS_CURSOR_Y are always the same, provide a
direction-independent synonym for them.

vadz committed:
Add wxSysMetricChangedEvent and generate it in wxMSW

This event can be used to react to the changes in the default system
cursor size.

vadz committed:
Construct wxCursor from wxBitmap in the image sample

There is no need to use an intermediate wxImage any more.

We also don't need to copy cursor.png to the build directory, as it is
now used only during the build and not when running the sample, so don't
do this.

vadz committed:
Use wxCursorBundle for testing cursor in wxImage sample

As part of this change, the cursor.png file of unclear provenance added
back in 9ea83ebc7c (Add test for a custom cursor to the image sample.,
2010-07-24) is replaced with the new hand image with a non-zero hotspot
point taken from https://uxwing.com/up-hand-pointer-icon/

vadz committed:
Implement support for wxCursorBundle in wxMSW

Update the cursor used by the window when the cursor size changes,
either due to a DPI change or to the change of the user-specified cursor
size.

vadz committed:
Fix comparing wxObjectDataPtr<> objects

Previously comparing objects of this type involved implicitly converting
both of them to the unspecified_bool_type and then comparing these
values, which meant that all non-null pointers were considered to be
equal.

Fix this now and add a trivial unit test which was failing before this
commit.

vadz committed:
Reset stored global cursor in wxMSW wxSetCursor() too

Previously, wxGetGlobalCursor() would still return the previously set
global cursor even after calling wxSetCursor(wxNullCursor) which was
unexpected and inconsistent with wxGTK.

Fix this by updating gs_globalCursor in any case, whether the new cursor
is valid or not.

vadz committed:
Add wxSetCursor() overload taking wxCursorBundle

Make it easier to set the cursor of the appropriate size.

vadz committed:
Avoid allocating wxMSW gs_globalCursor on the heap

This seems unnecessary, we can have global objects of wxCursor type as
destroying them doesn't really involve anything else.

This also allows to get rid of wxCursorModule which was only used to
initialize and clean up this global variable.

vadz committed:
Deprecate wxBusyCursorSuspender

This class only ever worked in wxGTK and wxX11 and doesn't seem really
useful: it's purported goal is to prevent busy cursor from being shown
for the modal dialogs, but nothing special needs to be done for this as
modal dialogs created after calling wxBeginBusyCursor() don't inherit
the busy cursor and so don't show it anyhow.

Just stop doing anything at all in this class and also in the functions
GetStoredCursor() and GetBusyCursor() that were only used by it.

vadz committed:
Extract busy cursor-related stuff into its own header

It didn't make sense to include wx/utils.h from wx/cursor.h, but it
somewhat does make sense to declare busy cursor-related functions and
wxBusyCursor itself in this header, so extract these declarations into
their own header which is included from both wx/cursor.h and wx/utils.h
for compatibility, but can be also included directly.

Document that wx/cursor.h doesn't pull in wx/utils.h any longer as it's
an incompatible change (but hopefully one not resulting in too much
disruption and so worth it).

vadz committed:
Make wxBusyCursor constructor explicit

It doesn't make sense to implicitly convert wxCursor pointers to
wxBusyCursor, so forbid it.

vadz committed:
Add wxCursorBundle::GetCursorForMainWindow()

This is just a helper calling GetCursorFor() with the application top
window.

vadz committed:
Provide wxBeginBusyCursor() overload taking wxCursorBundle

Make it simpler to use a high resolution busy cursor.

vadz committed:
Add wxCursorBundle ctor taking wxStockCursor

This allows the existing code passing wxStockCursor instead of wxCursor
parameter to keep working with the functions taking wxCursorBundle
instead of wxCursor too and, in particular, we can keep just a single
version of wxSetCursor() instead of having two overloads (which wasn't
necessarily a problem for just this function, but is inconsistent with
the other functions taking wxCursor, notably wxDropSource ctor where
overloads really can't be used because we would need too many of them
to cover all combinations of 3 different wxCursor parameters).

vadz committed:
Delegate to another ctor in wxDropSource implementations

Initializing wxDropSource with wxDataObject is identical to initializing
it without it and then calling SetData(): make this more clear by using
ctor delegation.

No real changes.

vadz committed:
Take wxCursorBundle in wxDropSource ctor instead of wxCursor

This allows to use high resolution cursors while dragging in the ports
that use cursors in wxDropSource, i.e. not wxGTK.

vadz committed:
Slightly improve behaviour of wxCursorBundle::GetCursorFor(null)

Check for the null window in the implementation class, which has 2
advantages:

1. We don't call wxApp::GetMainTopWindow() needlessly from
   GetCursorForMainWindow() if the implementation doesn't use the window
   anyhow (as is the case for a single-cursor bundle).
2. The code calling GetCursorFor() with a null window will get the
   cursor appropriate for the main window which is an improvement
   compared to always getting the system cursor size in this case.

vadz committed:
Allow using wxCursorBundle for wxHtmlWindow cursors

Extend SetDefaultHTMLCursor() to take wxCursorBundle and not just a
wxCursor and also take an (optional, for compatibility) wxWindow in
GetDefaultHTMLCursor() to return the cursor appropriate for this window.

vadz committed:
Replace wxHtmlWindowInterface::ms_cursorXXX with a function

This avoids the need to initialize these global variables on program
startup (which is cheap but still not worth doing if they are not,
perhaps, going to be ever used at all) and also allows to avoid
repeating the same code for the 3 cursor types as we can just use the
cursor type as an index instead.

vadz commented:
I think this should be good enough for now.

vadz commented:
@dsa-t I know that this doesn't leave a lot of time, but we plan to release 3.3.0 soon and I'd like to merge this a.s.a.p. to prepare for that release, so if you see any issues with this API, please let me know. TIA!

dsa-t commented:
Thanks for your work! I'll try it in KiCad today.

dsa-t commented:
There's at least one problem I see with `imagesample` on Linux Mint 21.1, X11, XFCE with window scaling 2x:

- `gtk-cursor-theme-size` is 24
- `wxSYS_CURSOR_Y` returns 48
- `m_bitmaps.GetPreferredBitmapSizeAtScale(double(48) / 32);`  returns 32
- So I get a 32x32 cursor that is scaled 2x by the system and looks blurry.

dsa-t commented:
(I think the theme cursor size shoudn't matter at that point (`wxCursorBundleMultiImpl::GetCursorFor`), only the display scaling factor)

dsa-t commented:
And I think `wxSYS_CURSOR_X`/`wxSYS_CURSOR_Y` shouldn't be scaled.

vadz commented:
I don't know why is the cursor getting scaled for you. I.e. I think we should be able to use 32x32 cursor as is...

dsa-t commented:
> I don't know why is the cursor getting scaled for you. I.e. I think we should be able to use 32x32 cursor as is...

GTK(?) does it automatically on my system, does it not on your system?

vadz commented:
It doesn't seem to... This is using Sway, I'll test under GNOME.

vadz commented:
Anyhow, the problem is that I'm really not sure what can we do if the cursor size is restricted and we don't have a bitmap with this size. Scaling it ourselves is not going to look (much) better than system-performed scaling.

vadz commented:
> It doesn't seem to... This is using Sway, I'll test under GNOME.

I take my words back, after viewing it with a magnifier (which I can only do in a VM, I wonder if there is some simpler way of viewing the cursor directly on the system), it is indeed doubled.

But this still doesn't help with understanding how can we prevent this from happening :-(

vadz commented:
At least under MSW the user can change the cursor size independently of the DPI and it's my understanding that we want to respect this setting, don't we? And if we do it for MSW it seems strange not to do it for GTK...

vadz commented:
I think something goes wrong with the scaling factor when creating the cursor, i.e. it's just not taken into account, so whichever bitmap we use, it's always doubled when DPI is 200%.

We might need to use `gdk_cursor_new_from_surface()` instead of `gdk_cursor_new_from_pixbuf()`...

dsa-t commented:
I think we can let the app manage this too, and in wx, only handle display scaling changes.

Or maybe we can let the app provide some `std::vector<wxBitmapBundle>` with different default sizes for different custom cursor sizes - but not sure it's worth it.

dsa-t commented:
> I think something goes wrong with the scaling factor when creating the cursor, i.e. it's just not taken into account, so whichever bitmap we use, it's always doubled when DPI is 200%.
> 
> We might need to use `gdk_cursor_new_from_surface()` instead of `gdk_cursor_new_from_pixbuf()`...

Check out the patch from https://github.com/wxWidgets/wxWidgets/issues/24564 -  it uses `cairo_surface_set_device_scale`

vadz commented:
> Check out the patch from #24564 - it uses `cairo_surface_set_device_scale`

Right, I've completely missed it there and was just writing something like this myself now. I'll push it here soon.

vadz committed:
Reset window cursor after it has been realized

The scale factor of the window is not known before it is realized, so
if we had already create wxCursor for it, we need to recreate it in case
the actual scale factor is different from 1.

vadz committed:
Take bitmap scale factor into account when creating wxCursor

We need to use gdk_cursor_new_from_surface() for creating the cursor in
order to be able to specify the bitmap scale factor (note that because
gdk_cursor_new_from_pixbuf() is implemented in terms of this function
anyhow, we don't gain anything from not calling it even if the scale
factor is 1).

This fixes unwanted cursor scaling by GTK/Wayland which made the cursor
both blurry and of a wrong size.

vadz committed:
Fix cursor bitmap size and scaling factor

Always use bitmap with the same scaling factor as the currently used one
but select the bitmap of the size corresponding to the scale of the
cursor (and not the default size of this cursor bundle itself).

This ensures that our cursor bitmaps are not rescaled and that we still
use a bigger bitmap even in standard DPI if the user increased the
default cursor size.

vadz commented:
I've spent an embarrassingly long time on deciding how to handle it and the result is in the last commit. I _think_ it does what we want, i.e. selects the best cursor size which makes sense for the current system taking into account both the DPI and the user-defined cursor size.

It feels very weird to me, however, to have to reset the scale factor manually like this, so it's quite possible that I'm still missing something... But it seems to work in my tests. What do you think?

dsa-t commented on src/gtk/cursor.cpp:
```diff
@@ -21,6 +21,7 @@
 #include "wx/gtk/private/wrapgtk.h"
```
> This fixes unwanted cursor scaling by GTK/Wayland which made the cursor
both blurry and of a wrong size.

Note that this is same on X11.

vadz replied:
True, but I had only tested this with Wayland. Anyhow, I don't think it's worth rewriting history/rerunning the CI once again, so I'll leave it as is.

dsa-t commented:
I still think the `wxSYS_CURSOR_X` / `wxSYS_CURSOR_Y` values shouldn't depend on display scaling.

Changes seem ok for me on wxMSW, though the default cursor switches to 2x on 150%+ scaling, while the wx cursor switches on 175%+.

On wxGTK I don't like the hardcoded values too much, but I guess that'll do it, unless GTK scales the cursor automatically to the custom size (not tested with latest commits).

vadz committed:
Prefer larger cursor in case of a tie between available sizes

Unlike for the bitmaps, using a larger cursor seems preferable to using
a smaller one in case of a tie (e.g. when using 150% DPI scaling with
only standard and @2x versions of the bitmap being available) and this
behaviour is consistent with what MSW does for the default cursor.

vadz commented:
> I still think the `wxSYS_CURSOR_X` / `wxSYS_CURSOR_Y` values shouldn't depend on display scaling.

But why? I mean, if the user selects 64x64 cursor under MSW, shouldn't we use the high DPI version instead of showing the blurry upscaled 32x32 one? AFAIK plenty of people increase their cursors size, so it makes sense to try to behave nicely in this case.

> Changes seem ok for me on wxMSW, though the default cursor switches to 2x on 150%+ scaling, while the wx cursor switches on 175%+.

Yes, this is because the default `GetPreferredBitmapSizeAtScale()` implementation prefers smaller size in case of a tie. I had initially decided to leave it like this, but the argument about the default cursor makes me think it's worth the extra ugliness of 11c3a2fc9cae5f3026654746492cb0d202faf72f.
 
> On wxGTK I don't like the hardcoded values too much, but I guess that'll do it, unless GTK scales the cursor automatically to the custom size (not tested with latest commits).

24 as default cursor size value comes from X11 and I don't think it's going to change in the observable future, so I'm not too worried. I'm more worried about other ports, e.g. wxQt now and maybe something else in the future, but I'm still not worried that much about this, and we can always virtualize this via platform-specific function if really needed.

dsa-t commented:
> > I still think the `wxSYS_CURSOR_X` / `wxSYS_CURSOR_Y` values shouldn't depend on display scaling.
> 
> But why? I mean, if the user selects 64x64 cursor under MSW, shouldn't we use the high DPI version instead of showing the blurry upscaled 32x32 one? AFAIK plenty of people increase their cursors size, so it makes sense to try to behave nicely in this case.

Other metrics in wxGTK don't depend on a window scaling factor, so I expect wxSYS_CURSOR_* to also not depend on it.

On MSW, e.g. `wxSYS_ICON_X` seems to change with display scaling, so it's probably fine.

vadz committed:
Return wxSYS_CURSOR_SIZE in logical pixels in wxGTK

This is consistent with the other metrics and wxOSX.

Update wxCursorBundle code to perform logical-to-physical translation.

vadz commented:
> Other metrics in wxGTK don't depend on a window scaling factor, so I expect wxSYS_CURSOR_* to also not depend on it.

Hmm, this is true, thanks for pointing this out.

> On MSW, e.g. `wxSYS_ICON_X` seems to change with display scaling, so it's probably fine.

So we have a choice between making this consistent with the other metrics under GTK or with the same metric under MSW. Looking at `src/osx/cocoa/settings.mm` I see that wxOSX also returns logical values, so I guess I'll have to conform to this, even if I don't really like it.

I've just pushed 901d31104af5b75d015d1f442d101a24b50af73a doing this.

Do you see any other remaining problems?

dsa-t commented:
Seems ok for me.
